### PR TITLE
Update to use platform interface and implement web, windows, macOS and split Linux into it's own project per best practices guidance from Google

### PR DESCRIPTION
**Pull request by jhancock4d**
Per: #252 here's the PR to merge this and implement web with full platform interface support. I also incidently fixed a bug in the iOS example project todo with the plugin name space. And I implemented macOs and windows options ready for those PRs to be merged (which could be their own projects as could the linux one now if you so desire to make them function that way as suggested by the flutter team for plugin development)

I have tested that I didn't break anything on Android, and tested the web functionality which appears to be done correctly.  It is my understanding from the referenced white papers that this results in the private key still be stored in the browser. I have tested this between browsers with the information that is stored in localstorage and I was not able to reverse the encryption on a second browser.

### CODE REVIEW

**Comment by juliansteenbakker** [comment]
First of all thank you very much for your work. This is definitely the way to implement this. 

I have just become maintainer of this repository and there are a few other open Pull requests that i'm gonna look to first. Also there are some other things like testing that i want to integrate so we are sure big changes like this don't break any functionality. Till then i have to test some of this myself so that will take some time. I do appreciate that you already tested it on Android and web.

**Comment by JohnGalt1717** [comment]
Thanks for the info!  The windows contributor has already agreed to add the remaining items for that to this pull request and the macOs pull request is already integrated.  If there are others that you'd like to see integrated I can look at them and setup the test harness stuff for you.

Pehaps this becomes a dev branch for the next version and all of the testing etc. gets done from this? It will be much easier to build on from here than the mono package going forward I would think as a result of the separation per google's design guides.

**Comment by jhancock4d** [comment]
Thanks for the info! The windows contributor has already agreed to add the remaining items for that to this pull request and the macOs pull request is already integrated. If there are others that you'd like to see integrated I can look at them and setup the test harness stuff for you.

Pehaps this becomes a dev branch for the next version and all of the testing etc. gets done from this? It will be much easier to build on from here than the mono package going forward I would think as a result of the separation per google's design guides.

File: example/.flutter-plugins-dependencies
Comment (juliansteenbakker) [inline example/.flutter-plugins-dependencies:1]: It seems like a lot of files have been checked in to VCS. Can you please check if all directories contain a correct .[gitignore](https://github.com/flutter/flutter/blob/master/.gitignore)?

**Review by juliansteenbakker — CHANGES_REQUESTED** [review]

**Comment by juliansteenbakker** [comment]
> Pehaps this becomes a dev branch for the next version and all of the testing etc. gets done from this.

This seems like a good idea but i first want to take a look at the older fixes that haven't been merged yet because otherwise they would all become invalid since the base dev branch changes. These seem like pretty small changes so i won't expect major merge conflicts.

**Comment by juliansteenbakker** [comment]
Also, which Pull request did you use for the macOS implementation? And which Windows pull request?

File: example/.flutter-plugins-dependencies
Comment (jhancock4d) [reply example/.flutter-plugins-dependencies]: Actually it was that there were old .gitignore files that weren't including all of the new things in flutter that needed to be ignored. I removed all of the old ones from the sub directories so that the root .gitignore could be used in a single location to make it easier in the future to update.  There are still ones in things like the ios and android folders per flutter's recommendations on this.

**Comment by jhancock4d** [comment]
macOs: https://github.com/mogol/flutter_secure_storage/pull/181 Looks like there are new changes in https://github.com/mogol/flutter_secure_storage/pull/255. Perhaps those can be integrated into this fork in the macOs project so that I don't mess it up?

Windows: https://github.com/mogol/flutter_secure_storage/pull/247

The contributor on #247 is going to use this to finish the last 2 methods for the Windows implementation from the comment in there. (readAll and deleteAll)

Also there appears to be pretty complete integration tests in the example project. If you don't object I'll update it to: https://flutter.dev/docs/testing/integration-tests styling.

I've stubbed unit tests innto the flutter_secure_storage project itself, and will fill them in too.  

From what I can gather here: https://github.com/flutter/flutter/issues/10824 doing proper platform unit tests isn't fully possible yet.

Although this gives a framework for doing so, not sure that it is actually viable in this case: https://github.com/flutter/flutter/wiki/Plugin-Tests

Thoughts?

**Comment by jhancock4d** [comment]
Note that I've added full unit tests for the platform interface in the mean time and force removed the one file that git took in that shouldn't be in there.

Let me know what else I can do to help get this in.

**Comment by juliansteenbakker** [comment]
Thank you for all the information. I will let you know as soon as i'm ready to test and merge this. I'm gonna see if i can get the automated tests working based on your information.

**Comment by jhancock4d** [comment]
@juliansteenbakker Note that the github actions needs to be changed to point to the right directories because of the change to the folder structure. I don't believe I have access to the actions to edit.

**Comment by juliansteenbakker** [comment]
> Note that the github actions needs to be changed to point to the right directories because of the change to the folder structure.
Thats no problem for now. I will change that myself.

File: example/.dart_tool/package_config_subset
Comment (matpag) [inline example/.dart_tool/package_config_subset:1]: `example/.dart_tool` folder should be gitignored

**Comment by juliansteenbakker** [comment]
@jhancock4d i am done with all older existing pull-requests. After we fix the current conflicts i will merge this in the develop branch and close the other redundant pull requests.

File: example/.dart_tool/package_config_subset
Comment (jhancock4d) [reply example/.dart_tool/package_config_subset]: Done.

**Comment by jhancock4d** [comment]
@juliansteenbakker for some reason this says that there are conflicts, but when I do a git merge upstream/develop I don't get any merges or anything.  Please advise.

**Comment by juliansteenbakker** [comment]
> @juliansteenbakker for some reason this says that there are conflicts, but when I do a git merge upstream/develop I don't get any merges or anything. Please advise.

Maybe stupid question, but did you execute git pull/git fetch on the develop branch?

**Comment by juliansteenbakker** [comment]
Also, when the merge conflicts are fixed, we should integrate #255 instead of the existing #181. I don't think this is alot of work since we can just replace the macOS implementation.

**Comment by jhancock4d** [comment]
Duh, forgot the pull first....


Ok, I'm getting errors with what I just checked in on android when running example. Looks like an encryption error. Could you take a look at the changes in this fork that were merged for the encrypted shared storage checkin? I believe that's what caused it but I can't figure out what the issue is.

(and I note that there is still a depreciation warning on build for android)

Will see about putting in the changes from #255. Not quite as straight forward so, it will need to be tested on a mac.

**Comment by jhancock4d** [comment]
@juliansteenbakker I believe that's it.  If you can check what's going on with Android that would be great and then I think we're good to go. Definately should, after this is put into develop, look at that android depreciation warning...

**Comment by jhancock4d** [comment]
PS: Updated the readme file. Might want to document the android secure storage thing but I don't know what it is all about.

And when #247 gets updated to include readAll and deleteAll from this once it's pulled into develop, both readmes should be updated accordingly.

**Comment by juliansteenbakker** [comment]
> Might want to document the android secure storage thing but I don't know what it is all about.

Yeah good you noticed.

>  If you can check what's going on with Android that would be great and then I think we're good to go. Definitely should, after this is put into develop, look at that android depreciation warning...

I will check this first and after that merge this. After this merge i'm gonna continue with the remaining tickets.

Thanks again for all of your hard work and the fast replies.

**Comment by jhancock4d** [comment]
Great! Let me know if there is anything else you need from me to pull this in to develop.

**Comment by juliansteenbakker** [comment]
I tested it on Android API 18 & 30 and it's working great! Unless @mogol has anything to say about it, i think it is ready to merge.

**Comment by juliansteenbakker** [comment]
While releasing v4.2.1 there were a few files that have been excluded from .gitignore. I think those can just be removed from this repository. Also, can you remove the .vscode folder and opt it in into the .gitignore? currently it is commented out:
```
# The .vscode folder contains launch configuration and tasks you configure in
# VS Code which you may wish to be included in version control, so this line
# is commented out by default.
#.vscode/
```

When this is done i will merge it.

**Comment by JohnGalt1717** [comment]
@juliansteenbakker Most projects are leaving the .vscode directory in because it provides easy testing and running and is common settings across machines. As an example, with that directory it becomes easy to run all of the tests directly from the debug screens in vs code. You might consider keeping it.

**Comment by jhancock4d** [comment]
@juliansteenbakker Most projects are leaving the .vscode directory in because it provides easy testing and running and is common settings across machines. As an example, with that directory it becomes easy to run all of the tests directly from the debug screens in vs code. You might consider keeping it.

**Pull request merged**
